crates/bounded_planet_core/src/networking/events.rs

```diff
@@ -63,12 +66,109 @@ pub enum ReceiveEvent {
-pub enum SendEvent {
+#[derive(Debug, Clone)]
+pub enum SendEvent
+{
     /// Send a packet through a specific stream on a specific connection
     SendPacket {
         connection: ConnectionId,
         stream: StreamType,
         data: Arc<Packet>,
     },
+
+    /// Send a packet through a new stream which is created just for this packet.
+    /// This should be used for very large packets.
+    TransferPacket {
+        connection: ConnectionId,
+        data: Arc<Packet>,
+    }
+}
+
+impl SendEvent {
+    pub fn get_connection(&self) -> ConnectionId {
+        match self {
+            SendEvent::SendPacket { connection, .. } => *connection,
+            SendEvent::TransferPacket { connection, .. } => *connection,
+        }
+    }
+
+    pub fn to_stream_sender_error(&self) -> NetworkError {
+        let (c, s, f) = match self {
+            SendEvent::SendPacket { connection, stream, data } => (*connection, Some(*stream), data),
+            SendEvent::TransferPacket { connection, data } => (*connection, None, data)
+        };
+
+        NetworkError::StreamSenderError {
+            connection: c,
+            stream: s,
+            failed_packet: f.clone()
+        }
+    }
+
+    pub async fn send(&self, senders: &mut Vec<SendStream<TlsSession>>, quinn_conn: &mut quinn::Connection) -> Result<(), NetworkError> {
+
+        match self {
+            SendEvent::SendPacket { stream, data, connection } => {
+                // Find (or create) the sender for this stream
+                let sender = match Self::get_stream_sender(&stream, senders, quinn_conn).await {
+                    Ok(sender) => sender,
+                    Err(err) => return Err(NetworkError::ConnectionError(err))
+                };
+
+                // Send the packet through the stream
+                if let Err(err) = data.send(sender).await {
+                    return Err(NetworkError::SendError {
+                        connection: *connection,
+                        stream: *stream,
+                        err,
+                    });
+                }
+
+                Ok(())
+            }
+
+            SendEvent::TransferPacket { data, .. } => {
+                // Open a new stream for this transfer
+                let uni = match quinn_conn.open_uni().await {
+                    Ok(uni) => uni,
+                    Err(err) => return Err(NetworkError::ConnectionError(err))
+                };
+
+                // Start sending the data
+                tokio::spawn(Self::send_transfer(uni, data.clone()));
+
+                Ok(())
+            }
+        }
+
+    }
+
+    async fn send_transfer(mut sender: SendStream<TlsSession>, data: Arc<Packet>) {
+        // Both of these method generate a result which is discarded.
+        // Results from transfers are not sent anywhere as that could potentially result in
+        // errors from a connection arriving in the ECS after is has closed!
+        let _ = data.send(&mut sender).await;
```

Comment:
I can't tell if the comment is explaining why this is ok, or why it's a problem...

Reply:
well... kinda both. I'm not entirely happy about these transfer errors being lost, but I'm also not sure it's a massive error as the client can just re-request what it needs if it never arrives.